### ISSUE TITLE
Provide examples

### ISSUE THREAD

**Issue by MCUdude** [nerdralph/picoI2C#2]
Hi Ralph!

I was thinking about adding a software i2c library to [MicroCore](https://github.com/MCUdude/MicroCore), and this library looks like it may be a good match. It would be really cool to provide an ultra light-weight, bit-banged i2c library for the ATtiny13, so I hope you have time to help me get started.

Even though I haven't tried it on actual hardware yet (I'm currently on vacation with no hardware in my luggage) it's a little bit difficult to figure out what the library can and can't do.

* Does this library support slave mode, or is it master only?
* Would there be any way to set the i2c clock speed, by frequency or clock divider?
* Let's say I want to check if a device with address 0x20 is present. What does a scan sequence look like? The provided example is kinda incomplete.
* What would a "blind" write example look like? Perhaps this?
  ```c++
  PicoI2C::startWrite(0x20); // i2c address 0x20
  PicoI2C::write(0x00); // First byte
  PicoI2C::write(0x01); // Second byte
  PicoI2C::write(0x10); // Third byte
  PicoI2C::stop();
  ```
* How can I read N bytes from an i2c slave? With the standard Arduino Wire library it can be done like this:
  ```c++
  Wire.beginTransmission(0x20); // i2c address 0x20
  Wire.write(0x01); // We want to read register 0x01
  Wire.endTransmission();
  
  Wire.requestFrom(0x20, 1); // Request/Read one byte from address 0x20
  while (Wire.available())
  {
    char c = Wire.read(); // receive a byte as character
    Serial.print(c);         // print the character
  }

  ```

Thanks!

**Comment by nerdralph** [comment]
I never came up with a satisfactory way to implement slave mode.  It certainly can be done, but the hard part is keeping it simple with minimal code size, while still properly recognizing start & stop conditions.  I also tried to keep the interface somewhat similar to the Arduino Wire class, even though I think the Wire class is a poor abstraction for I2C communications.  I left it as a work in progress last year, with the intention of coming back to it when I came up with some better ideas.  I agree it could use a better I2C scanner.  The library requires the SDA and SCL pins to be known at compile time, but for a scanner it would be nice if they could change at runtime so the scanner could detect things like swapped SDA and SCL.

I haven't had much time for coding over the last few months, and I'll probably be busy with other priorities until late fall.  It is still something I'd like to put more work into.  I have an old SDRAM DIMM with a I2C eeprom for SPD info sitting on my desk that I've thought about using to test the library, and I also have an I2C EEPROM module from an old printer cartridge that I got partially working with the library when I was initially developing it.

**Comment by MCUdude** [comment]
> I never came up with a satisfactory way to implement slave mode. It certainly can be done, but the hard part is keeping it simple with minimal code size, while still properly recognizing start & stop conditions.

Slave mode would be interesting to have to work on an ATtiny13. I see many use cases where the t13 can read a few sensors and communicate with a master device over i2c.

> I also tried to keep the interface somewhat similar to the Arduino Wire class, even though I think the Wire class is a poor abstraction for I2C communications. I left it as a work in progress last year, with the intention of coming back to it when I came up with some better ideas

Even though it might be nice to imitate the Wire class, I don't think it's very much of a deal-breaker if the ATtiny13 library has a slightly different syntax. Everything Arduino-ATtiny13 is pretty niche anyways, so I don't users would demand full portability between larger Arduino boards and the ATtiny13. I could of course rename the functions to match the Wire library, but the underlying functionality would be slightly different anyway.

> The library requires the SDA and SCL pins to be known at compile time, but for a scanner it would be nice if they could change at runtime so the scanner could detect things like swapped SDA and SCL.

Since the ATtiny13's memory is limited, I'm totally fine with that SDA and SCL have to be known at compile time. If the library turns to be working in its current state, I could implement it as a master-only i2c library, and maybe add slave functionality later if possible.

> I haven't had much time for coding over the last few months, and I'll probably be busy with other priorities until late fall. It is still something I'd like to put more work into. I have an old SDRAM DIMM with a I2C eeprom for SPD info sitting on my desk that I've thought about using to test the library, and I also have an I2C EEPROM module from an old printer cartridge that I got partially working with the library when I was initially developing it.

Please let me know when you have time and desire to continue working on the library! I'll gladly help out with testing and troubleshooting. A lightweight, efficient, and easy-to-use i2c library for the ATtiny13 could lead to some very interesting projects!

**Comment by MCUdude** [comment]
I've done some more C++ investigation, and it seems that having all function members in the class static reduces the compiled size by about 300 bytes! @nerdralph do you know why a class with static members (and thus no object in use) compiles down to a vastly smaller program with avr-gcc? Is there a compiler trick or feature here that I'm not aware of?

**Comment by nerdralph** [comment]
> 
> 
> I've done some more C++ investigation, and it seems that having all function members in the class static reduces the compiled size by about 300 bytes! @nerdralph do you know why a class with static members (and thus no object in use) compiles down to a vastly smaller program with avr-gcc? Is there a compiler trick or feature here that I'm not aware of?

Yes, I noticed this a few years ago, and used a static class with https://github.com/nerdralph/TM1638NR
Global static object instances is one of the worst coding practices that I see in Arduino libraries.  clang/lld does a better job than gcc/LTO at optimizing away singletons, but lld hasn't been been ported to AVR.

**Comment by MCUdude** [comment]
I've now pushed a test version of my PicoI2C implementation. It pretty much is the exact library you've written, but I've renamed some of the functions so that it's more familiar for the average Arduino user. The library still has static members, but I've created an object so the familiar `Wire.` syntax can be used.

I've provided an i2c scanner sketch and a sketch where one can read and write to an i2c eeprom.

https://github.com/MCUdude/MicroCore/tree/softwareI2C/avr/libraries/TinyWire

There are the features I hope to add later

* More examples
* Optional internal pullups (if possible, I have yet to get it working)
* Adjustable clock speed
  - The clock is currently running at 420kHz at 8 MHz. Dividing this down by 2, 4, or 8 somehow would be really nice!
* I2C slave support

Thank you for your excellent work so far! It's pretty incredible that we can have a working i2c library on the Attiny13!
